### DISCUSSION TITLE
What's the best way to change the default keyboard font size through code?

### DISCUSSION THREAD

**Discussion by sabzo** [florisboard/florisboard#1385] [Q&A]
What's the best way to change the default keyboard font size through code?

It seems changing the `fontSizeMultiplier`, `labelPaintTextSize`, within _TextKeyboardView.kt_ doesn't help. This affects low-budget devices that don't have a high screen resolution and makes the text look too small (with the exception of text when in All caps mode). 

@patrickgold

**Comment by sabzo** [comment]
Also, changing the font on `Type.kt` only affects the text on the preferences screen dropdown section (the one about giving feedback on the new design)

```
val Typography = Typography(
    body1 = TextStyle(
        fontFamily = FontFamily.Default,
        fontWeight = FontWeight.Normal,
        fontSize = 20.sp
    )
```

**Reply by patrickgold** [reply]
Actually I will use this thread for the Compose discussion.

In general I would recommend the official guide on JC for explaining the basics how Compose works: https://developer.android.com/jetpack/compose/mental-model 

This guide helped me a lot understanding the basics, and in JC quite a few things work differently than in the traditional view-based UI system.

So in general `MaterialTheme.typography` is a read-only composition local, which summarized is a fancy safe global variable only available within a composable context. Each composition local provides a value previously defined by the parent scope (may be any parent, the first parent in-to-out has the say what the value for the inner scope is).

This means that both a library and the app itself can call `MaterialTheme.typography` and will always be able to retrieve the current value set by the composition context. So no more `resources.getDimension(R.dimen.text_size)`, but simply MaterialTheme.typography.body1` for instance. The cool thing is that this default API supports the material color set (`MaterialTheme.colors.*`), shapes and typography. This is mostly enough for each general UI library to theme each UI element correctly, without manually needing to assign each material property to each composable (view in the old model).

In the case of JetPref and your above code snippet, JetPref currently does not take advantage of this global `MaterialTheme.typography` concept, so each text style (and thus font size too) falls back to `TextStyle.Default`, which provides the default font size of your device. Thus, changing body1 will not affect the preference text at all atm.

I need to change the implementation of the inner `JetPrefListItem` to give the title and secondary text style and get the style from the current `MaterialTheme`. Will link the relevant fix here once I push the commit for it.

**Reply by patrickgold** [reply]
Ok actually the list item already takes advantage of this system, just hidden within the default list item implementation from JC:

```kt
    val typography = MaterialTheme.typography

    val styledText = applyTextStyle(typography.subtitle1, ContentAlpha.high, text)!!
    val styledSecondaryText = applyTextStyle(typography.body2, ContentAlpha.medium, secondaryText)
    val styledOverlineText = applyTextStyle(typography.overline, ContentAlpha.high, overlineText)
    val styledTrailing = applyTextStyle(typography.caption, ContentAlpha.high, trailing)
```

This means the title would need `MaterialTheme.typography.subtitle1` and the secondary text `MAterialTheme.typography.body2` to modify the text size of those.

Buttons use the text style from `button`, and radio list items are `body1`. The screen title in the app bar is `h6`.

Note that what control what style question is solved by the JC team, so I think this should be correct in terms of material theme style compliance.

**Comment by patrickgold** [comment]
> fontSizeMultiplier

This affects the hinted font size, key font size, but NOT the popup font size (which I just discovered and which is absolutely weird/should not be that way).

In general the font size is somewhat hard-coded (I know I know if I had written FlorisBoard from scratch with my knowledge and experience from today I would know that this is horrible, but at the time this was fine for me). This is actually one thing that the new super-experimental internal theme engine will provide: each element which has the ability to show text or an icon will have the ability to change the `font-size` attribute. But that won't help you now sadly :(

If you want to change the icon/font size, you probably need to adjust the popup view, the Smartbar quick action view, the clipboard view, the emoji view. For the normal keys, adjusting the fontSizeMultiplier should be sufficient. If you want I can go search where I hard-coded the font sizes, I don't know them by memory currently.

> Also, changing the font on Type.kt only affects the text on the preferences screen dropdown section (the one about giving feedback on the new design)

This is actually mainly an issue of the JetPref UI library (https://github.com/patrickgold/jetpref), which provides the base implementation for Compose Preferences and ignores the Floris MaterialTheme typography. If you want you can have the honor to open the first issue there and we handle this specific issue in the JetPref repo, or we can discuss it in this thread. Important is that you are right and it is an issue which needs addressing.

Note: currently Compose only affect the preference UI, not the keyboard UI. So that's not a bug, just to clarify.

**Reply by sabzo** [reply]
> This is actually mainly an issue of the JetPref UI library (https://github.com/patrickgold/jetpref), which provides the base implementation for Compose Preferences and ignores the Floris MaterialTheme typography. If you want you can have the honor to open the first issue there and we handle this specific issue in the JetPref repo, or we can discuss it in this thread. Important is that you are right and it is an issue which needs addressing.

I wasn't sure if that was a bug or not as I'm not familiar with the new Jetpack way of doing things. But I suppose we can discuss it here as I'm new to it. 

> If you want I can go search where I hard-coded the font sizes, I don't know them by memory currently.

Yes if you may, that would be helpful.
> This is actually one thing that the new super-experimental internal theme engine will provide: each element which has the ability to show text or an icon will have the ability to change the font-size attribute.

Which release would this be?

**Reply by patrickgold** [reply]
> Yes if you may, that would be helpful.

Ok I found out that not everything is hard-coded. If you change these values, you should come very far: https://github.com/florisboard/florisboard/blob/bce5199e311d7048ffad14d7421b18a8f9997d24/app/src/main/res/values/dimens.xml#L24-L30

Clipboard: adding a text size attribute to this view: https://github.com/florisboard/florisboard/blob/bce5199e311d7048ffad14d7421b18a8f9997d24/app/src/main/res/layout/clipboard_history_item_text.xml#L13-L23

And here: https://github.com/florisboard/florisboard/blob/bce5199e311d7048ffad14d7421b18a8f9997d24/app/src/main/res/layout/clipboard_layout.xml#L31-L38

The smartbar quick action is trickier, because it auto scales to the space it has. The style for each quick action is defined here: https://github.com/florisboard/florisboard/blob/bce5199e311d7048ffad14d7421b18a8f9997d24/app/src/main/res/values/styles.xml#L11-L20

Let me know if this is enough or if I forgot some size controls.

> I wasn't sure if that was a bug or not as I'm not familiar with the new Jetpack way of doing things. But I suppose we can discuss it here as I'm new to it.

I will start a new thread below it (same discussion though) to separate this topic a bit)

> Which release would this be?

No clue, probably beta08 will have the first preview for the Smartbar only, then other keyboard UI controls will slowly follow.
